Restrict app summary views to users with at least view apps permissions

gherceg opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->

## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
https://dimagi.atlassian.net/browse/SAAS-18270

Prevent users without at least the view apps permission from viewing the following views:
- AppDataView
- AppSummaryView
- DownloadAppSummaryView
- DownloadCaseSummaryView

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
This is just enforcing an existing permission we already have defined in HQ. The risk would be if there is a legitimate reason that a user without the view apps permission would need to view an app summary page, but to me this change feels in line with what this permission should and should not allow access to.

Tested locally and on staging.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
No tests

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
No QA

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
Put app summary behind view apps privilege

gherceg committed:
Lint errors

gherceg committed:
Organize imports

orangejenny reviewed (approved):
Looks fine.

Alternative: could it make sense to incorporate this check into `ApplicationViewMixin`?

gherceg committed:
Revert one-off permissions check

gherceg committed:
Add require_can_edit_or_view_apps to mixin

It makes sense for the ApplicationViewMixin to have a permission check
for being able to view apps.

gherceg committed:
Call ApplicationViewMixin's dispatch before view

If the view's dispatch is called before a mixin, it is unlikely to pass
it along since it implements its own dispatch and does not call
super().dispatch

gherceg committed:
Replace can_view_or_edit with can_view

This decorator isn't being used anywhere

gherceg committed:
Add permission override to relevant tests

gherceg committed:
Move login_or_api_key to dispatch

gherceg committed:
Only allow get requests

gherceg committed:
Merge branch 'master' into gh/bac/applications

gherceg commented on corehq/apps/app_manager/views/app_summary.py:
```diff
@@ -486,7 +485,11 @@ class DownloadCaseSummaryView(ApplicationViewMixin, View):
     urlname = 'download_case_summary'
     http_method_names = ['get']
 
+    @method_decorator(require_GET)
```
I wouldn't have added this but did because of an existing test that enforced that a 405 would be returned for any http method besides GET. Previously this wasn't needed because this view didn't implement a `dispatch`, and therefore the View.dispatch method could check which http methods the summary view supports, and raise a method not allowed error for methods it does not support.

gherceg commented on corehq/apps/hqmedia/views.py:
```diff
@@ -900,7 +900,7 @@ def iter_app_files(app, include_multimedia_files, include_index_files,
 
 
-class DownloadMultimediaZip(View, ApplicationViewMixin):
+class DownloadMultimediaZip(ApplicationViewMixin, View):
```
The fact that we don't currently have any login decorator on this view certainly indicates that.

gherceg committed:
Revert require_GET change

gherceg committed:
Explicitly set permissions on views that need it

millerdev commented on corehq/apps/app_manager/views/app_summary.py:
```diff
@@ -175,6 +176,7 @@ class AppDataView(LoginAndDomainMixin, ApplicationViewMixin, View):
 
     urlname = 'app_data_json'
 
+    @method_decorator(require_can_edit_or_view_apps, name='dispatch')
     def get(self, request, *args, **kwargs):
```
I didn't know you could use `method_decorator` on a method to decorate a different method. I assume you've verified that this works.

millerdev reviewed (approved):
Approval conditional on my comment not highlighting a typo.

gherceg committed:
Don't specify dispatch on a get method decorator
